### PR TITLE
Packet Obfuscation support

### PR DESCRIPTION
**Pull request by cydh**
* Enable it by uncomment `#define PACKET_OBFUSCATION` in mmo.h
* Obfuscation keys are defined in db/packet_db.txt, each packet_ver has their own default keys (for temperory lookup) in `packet_keys:` right after `packet_ver:`
* Keys can be refined (user-defined keys) by defining the keys in `packet_keys_use:` in file db/[import/]packet_db.txt
* This feature only for client 20110817 or newer
* `@reloadpacketdb` won't changes the keys, so make sure define everything before run the map-server.

Thank to @icxbb-xx, @aleos, https://github.com/HerculesWS/Hercules/commit/5a942993dd12ae7d85b1633a58b0dc23f85832c2

Signed-off-by: Cydh Ramdh <house.bad@gmail.com>

### CODE REVIEW

**Comment by cydh** [comment]
forgot to tell, I use `paket_keys` format from Lilystar (see [post](https://rathena.org/board/topic/99006-2014-10-22-client-roulette-and-per-char-gender/#entry275722) [svn](https://subversion.assembla.com/svn/translation-j/packet_lengths/rAthena/) )

**Comment by RadianFord** [comment]
Wow really? this is a good news :+1:

**Comment by aleos89** [comment]
@RadianFord: Essentially you don't want to apply the diff to remove encrypted packets.

**Comment by RadianFord** [comment]
@aleos89 but i can apply this packet obfuscation right?

**Comment by RadianFord** [comment]
@aleos89 Oh okay, I will still test this one too. thanks!

**Comment by julia40124009** [comment]
good job!

**Comment by RadianFord** [comment]
Warning after i disable it on the core.h

clif.c: In function ‘packetdb_readdb’:
clif.c:17779: warning: unused variable ‘last_key_defined’
clif.c:17776: warning: unused variable ‘key_defined’

File: src/map/clif.c [hunk 1]
```diff
@@ -17527,6 +17536,26 @@ void clif_party_leaderchanged(struct map_session_data *sd, int prev_leader_aid,
...
+ * Decrypt packet identifier for player
+ * @param fd
+ * @param sd
+ * @param packet_ver
+ * Orig author [Ind/Hercules]
+ **/
+static unsigned short clif_parse_cmd(int fd, struct map_session_data *sd) {
+#ifndef PACKET_OBFUSCATION
+	return RFIFOW(fd, 0);
+#else
```
Comment (Lemongrass3110) [inline src/map/clif.c]: Mind to add some comments here why this things are done?
I will add some suggestions

File: src/map/clif.c [hunk 2]
```diff
@@ -17527,6 +17536,26 @@ void clif_party_leaderchanged(struct map_session_data *sd, int prev_leader_aid,
...
+	unsigned short cmd = RFIFOW(fd,0);
```
Comment (Lemongrass3110) [inline src/map/clif.c]: // Check if it is a player that tries to connect to the map server

File: src/map/clif.c [hunk 3]
```diff
@@ -17527,6 +17536,26 @@ void clif_party_leaderchanged(struct map_session_data *sd, int prev_leader_aid,
...
+	if (sd)
```
Comment (Lemongrass3110) [inline src/map/clif.c]: // Decrypt the current packet ID with the last key stored in the session

File: src/map/clif.c [hunk 4]
```diff
@@ -17527,6 +17536,26 @@ void clif_party_leaderchanged(struct map_session_data *sd, int prev_leader_aid,
...
+		cmd = (cmd ^ ((sd->cryptKey >> 16) & 0x7FFF));
+	else
```
Comment (Lemongrass3110) [inline src/map/clif.c]: // A player tries to connect - use the initial keys for the decryption of the packet ID

File: src/map/clif.c [hunk 5]
```diff
@@ -17527,6 +17536,26 @@ void clif_party_leaderchanged(struct map_session_data *sd, int prev_leader_aid,
...
+		cmd = (cmd ^ ((((clif_cryptKey[0] * clif_cryptKey[1]) + clif_cryptKey[2]) >> 16) & 0x7FFF));
```
Comment (Lemongrass3110) [inline src/map/clif.c]: // Return the decrypted packet ID

File: src/map/clif.c [hunk 6]
```diff
@@ -18265,6 +18308,36 @@ void packetdb_readdb(void)
...
+#ifdef PACKET_OBFUSCATION
+				else if (!reload && strcmpi(w1,"packet_keys") == 0) {
+					char key1[12] = { 0 }, key2[12] = { 0 }, key3[12] = { 0 };
+					trim(w2);
+					if (sscanf(w2, "%11[^,],%11[^,],%11[^ \r\n/]", key1, key2, key3) == 3) {
+						CREATE(packet_keys[packet_ver], struct s_packet_keys, 1);
+						packet_keys[packet_ver]->keys[0] = strtol(key1, NULL, 0);
+						packet_keys[packet_ver]->keys[1] = strtol(key2, NULL, 0);
+						packet_keys[packet_ver]->keys[2] = strtol(key3, NULL, 0);
+						last_key_defined = packet_ver;
```
Comment (Lemongrass3110) [inline src/map/clif.c]: Why do we store the keys for all packet versions here but dont use them depending on the packet version in the first call to the decrypt function?

**Pull request merged**